[WPE] Add support for touch based pointer events

Bastian-Krause opened the pull request:
#### 84e4a4ae1e8936da66c30862be411246417d44fd
<pre>
[WPE] Add support for touch based pointer events
<a href="https://bugs.webkit.org/show_bug.cgi?id=214870">https://bugs.webkit.org/show_bug.cgi?id=214870</a>

Reviewed by Carlos Alberto Lopez Perez and Michael Catanzaro.

Currently all non-Apple ports support only pointer events of type
&quot;mouse&quot; because they a mapped 1:1 to mouse events. This commit adds
the pointer event support for touch events for the WPE port. With
this patch it should be easy for other ports to add pointer event support
as well.

The WPE port creates artificial motion, button down, button up events
for short touch gestures in PageClientImpl::doneWithTouchEvent() to keep
compatibility, see:

<a href="https://w3c.github.io/pointerevents/#compatibility-mapping-with-mouse-events">https://w3c.github.io/pointerevents/#compatibility-mapping-with-mouse-events</a>

These NativeWebMouseEvents are undistinguishable from actual mouse
events and lead to additional pointer events of type mouse being
dispatched.
Fix this by passing the SyntheticClickType down to the WebMouseEvent and
PlatformMouseEvent. With this, WPE can make use of the code path
introduced in <a href="https://bugs.webkit.org/show_bug.cgi?id=205551">https://bugs.webkit.org/show_bug.cgi?id=205551</a> due to the
same issue.

Initial patch version by Marco Felsch &lt;m.felsch@pengutronix.de&gt;. Tested
and improved by Bastian Krause &lt;bst@pengutronix.de&gt;.

* Source/WebCore/SourcesWPE.txt: Add dom/wpe/PointerEventWPE.cpp, see
below.
* Source/WebCore/dom/PointerEvent.h: Enable various pointer event paths
for WPE, move buttonForType() and buttonsForType() generics from
PointerEventIOS.
* Source/WebCore/dom/wpe/PointerEventWPE.cpp: Adapted for WPE from
Source/WebCore/dom/ios/PointerEventIOS.cpp.
(WebCore::pointerEventType): Add a translation function similar to
PointerEventIOS&apos;s pointerEventType().
(WebCore::PointerEvent::create): Add create functions similar to
PointerEventIOS.
(WebCore::PointerEvent::PointerEvent):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleTouchEvent): Change point iteration to
make index value accessible. Add pointer target hit tests for
TouchReleased/TouchCancelled/TouchMoved. Implement WPE-specific
workaround to compensate for non-existent touch stationary events.
Finally dispatch pointer events.
* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::cancelPointer): Enable existing
pointer event code paths for WPE.
* Source/WebCore/page/PointerCaptureController.h:
(WebCore::PointerCaptureController::CapturingData::hasAnyElement const):
* Source/WebCore/platform/PlatformTouchEvent.h:
(WebCore::PlatformTouchEvent::setTouchPoints): Allow setting touchPoints
for WPE to make the above mentioned touch stationary worarkound
possible.
* Source/WebCore/platform/PlatformTouchPoint.h:
(WebCore::PlatformTouchPoint::PlatformTouchPoint): Allow WPE to create
PlatformTouchPoints to make the above mentioned touch stationary
worarkound possible.
* Source/WebKit/Shared/NativeWebMouseEvent.h: Add synthetic click type.
* Source/WebKit/Shared/WebEventConversion.cpp:
(WebKit::WebKit2PlatformMouseEvent::WebKit2PlatformMouseEvent): Pass the
SyntheticClickType to the PlatformMouseEvent, for WPE only because effects
on other ports are unknown at this time.
* Source/WebKit/Shared/libwpe/NativeWebMouseEventLibWPE.cpp:
(WebKit::NativeWebMouseEvent::NativeWebMouseEvent): Add synthetic click
type.
* Source/WebKit/Shared/libwpe/WebEventFactory.cpp: Add synthetic click
type.
(WebKit::WebEventFactory::createWebMouseEvent):
* Source/WebKit/Shared/libwpe/WebEventFactory.h: Pass synthetic click
type to WebMouseEvent.
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::doneWithTouchEvent): Set synthetic click type
for synthetic mouse events to NativeWebMouseEvent.

Canonical link: <a href="https://commits.webkit.org/263969@main">https://commits.webkit.org/263969@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fcef3c978b920c14e6c5a19e6d48056547134a4f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6237 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6430 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6609 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7806 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6568 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6237 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6650 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6380 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9448 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6348 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6363 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5648 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7896 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3836 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5627 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/13536 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5701 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5707 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7973 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6185 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5057 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5597 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1484 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9745 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5964 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shivamidow commented on Source/WebCore/dom/PointerEvent.h:
```diff
@@ -116,13 +119,24 @@ class PointerEvent final : public MouseEvent {
 #if ENABLE(TOUCH_EVENTS) && PLATFORM(IOS_FAMILY)
     PointerEvent(const AtomString& type, const PlatformTouchEvent&, IsCancelable isCancelable, unsigned touchIndex, bool isPrimary, Ref<WindowProxy>&&);
+#elif ENABLE(TOUCH_EVENTS) && PLATFORM(WPE)
+    PointerEvent(const AtomString&, Ref<WindowProxy>&&, const Touch&, float, IsPrimary);
```
*This is an unofficial review. Just refer.*

For a primitive argument, we can use a variable name in method declarations. `float pressure`

Bastian-Krause replied:
Thanks, fixed.

shivamidow commented on Source/WebCore/dom/wpe/PointerEventWPE.cpp:
```diff
@@ -0,0 +1,58 @@
+Ref<PointerEvent> PointerEvent::create(const AtomString& type, Ref<WindowProxy>&& view, const Touch& touch, IsPrimary isPrimary)
+{
+    // FIXME: Pressure is not supported currently. In this case the pressure should be set to 0.5 when in
+    // 'active buttons state' and to 0 otherwise according the specification:
```
*This is an unofficial review. Just refer.*

Maybe, what you want to say is.
```
// Where the pressure is not supported, it must be 0.5 in the active button state—otherwise, 0.
// https://www.w3.org/TR/pointerevents/#pointerevent-interface
```

Bastian-Krause replied:
No, as you can see, we don't pass an actual pressure value here, but rather use 0 and 0.5 as specified for platforms that do not support pressure:
> For hardware and platforms that do not support pressure, the value MUST be 0.5 when in the [active buttons state](https://www.w3.org/TR/pointerevents/#dfn-active-buttons-state) and 0 otherwise. Note: all pointerup events will have pressure 0.

Source: https://www.w3.org/TR/pointerevents/#dom-pointerevent-pressure

Actual pressure handling can be added later. Most touch controllers and drivers do not have this information anyway.

Bastian-Krause replied:
This is now reworked completely, please have a look at the recent version.

shivamidow commented on Source/WebCore/dom/wpe/PointerEventWPE.cpp:
```diff
@@ -0,0 +1,58 @@
+    // 'active buttons state' and to 0 otherwise according the specification:
+    // https://www.w3.org/TR/pointerevents/#pointerevent-interface
+    float pressure = 0.5;
+    if (type == eventNames().pointerupEvent || type == eventNames().pointeroutEvent || type == eventNames().pointerleaveEvent)
```
*This is an unofficial review. Just refer.*

So... by checking these three-pointer events, can we decide if the hardware is capable of detecting pressures?

Bastian-Krause replied:
No, see above.

Bastian-Krause replied:
This has been reworked completely. No need for any of this anymore. Please have a look at the recent version.

shivamidow commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4722,6 +4744,146 @@ static HitTestResult hitTestResultInFrame(Frame* frame, const LayoutPoint& point
+#if PLATFORM(WPE)
+// Mostly copied from handleTouchEvent() since the most notable change is the requirement
+// to perfom the hit test on every event.
+void EventHandler::handleTouchPointerEvent(const PlatformTouchEvent& event)
```
*This is an unofficial review. Just refer.*

touchPointerEvent sounds confusing. handlePointerEventWithTouch()?

Bastian-Krause replied:
I'd like to wait for maintainer feedback if this should be renamed.

Bastian-Krause replied:
This has been reworked completely. No need for any of this anymore. Please have a look at the recent version.

shivamidow commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4722,6 +4744,146 @@ static HitTestResult hitTestResultInFrame(Frame* frame, const LayoutPoint& point
 
+#if PLATFORM(WPE)
+// Mostly copied from handleTouchEvent() since the most notable change is the requirement
+// to perfom the hit test on every event.
```
*This is an unofficial review. Just refer.*

If that is the case, what about putting the most notable change with if-end guards inside handleTouchEvent()?

Bastian-Krause replied:
I don't think that's a good idea. `handleTouchPointerEvent()` is called from multiple locations and the new function behaves differently in some key aspects. Mixing these will make handleTouchEvent() huge and the if-end guards make it unnecessarily confusing.

Bastian-Krause replied:
The recent rework made it indeed possible to get rid of the extra function. Please have a look at the recent version.

shivamidow commented on Source/WebCore/page/PointerCaptureController.cpp:
```diff
@@ -310,7 +330,28 @@ void PointerCaptureController::dispatchEventForTouchAtIndex(EventTarget& target,
+}
+#endif
+
+#if PLATFORM(WPE)
```
*This is an unofficial review. Just refer.*

#elif PLATFORM(WPE)

Bastian-Krause replied:
Thanks, fixed.

shivamidow commented on Source/WebCore/page/PointerCaptureController.h:
```diff
@@ -59,6 +61,8 @@ class PointerCaptureController {
 #if ENABLE(TOUCH_EVENTS) && PLATFORM(IOS_FAMILY)
     void dispatchEventForTouchAtIndex(EventTarget&, const PlatformTouchEvent&, unsigned, bool isPrimary, WindowProxy&);
+#elif ENABLE(TOUCH_EVENTS) && PLATFORM(WPE)
+    void dispatchEventForTouchList(const AtomString&, const TouchList&, WindowProxy&, bool = false);
```
*This is an unofficial review. Just refer.*

bool isPrimary?

Bastian-Krause replied:
Thanks, fixed.

shivamidow commented on Source/WebCore/dom/wpe/PointerEventWPE.cpp:
```diff
@@ -0,0 +1,58 @@
+#include "config.h"
+#include "PointerEvent.h"
+
+#if ENABLE(TOUCH_EVENTS) && PLATFORM(WPE)
```
*This is an unofficial review. Just refer.*

The WPE platform uses this file only. So, I think PLATFORM(WPE) is redundant.

Bastian-Krause replied:
True, I was confused by the` PLATFORM(IOS)` in the IOS counterpart of this file. I just realized that this is necessary there because only a single `Source/WebCore/SourcesCocoa.txt` exists for Mac/IOS. I've dropped it here.

shivamidow commented on Source/WebCore/page/PointerCaptureController.cpp:
```diff
@@ -195,7 +201,21 @@ bool PointerCaptureController::preventsCompatibilityMouseEventsForIdentifier(Poi
+    UNUSED_PARAM(platformTouchEvent);
+    UNUSED_PARAM(index);
+    UNUSED_PARAM(isPrimary);
+
```
*This is an unofficial review. Just refer.*

why don't we have an exception handling for the empty type like line 208?

Bastian-Krause replied:
The empty type happens only for IOS because `PointerCaptureController::dispatchEventForTouchAtIndex()` is called from there omitting the type argument. All other calls pass a type explicitly.

I've added a `ASSERT(!type.isEmpty())` here.

clopez commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4762,7 +4763,8 @@ bool EventHandler::handleTouchEvent(const PlatformTouchEvent& event)
             allTouchReleased = false;
     }
 
-    for (auto& point : points) {
+    for (unsigned index = 0; index < points.size(); index++) {
+        auto& point = points[index];
         PlatformTouchPoint::State pointState = point.state();
         LayoutPoint pagePoint = documentPointForWindowPoint(m_frame, point.pos());
```
why is this change needed?

Bastian-Krause replied:
Quote from the commit message:
> (WebCore::EventHandler::handleTouchEvent): Change point iteration to
make index value accessible. [...]

The `index` is used down in the `pointerCaptureController.dispatchEventForTouchAtIndex()` call.

clopez replied:
I see. I missed that, sorry.

mcatanzaro reviewed (approved):
I'm not sufficiently familiar with event handling to spot bugs, but this seems generally OK.

So currently only iOS and WPE will support touch pointer events? This means GTK will not, even though it has had touch support for a long time?

clopez reviewed (changes requested):
Please upload a new patch with the above fix or a similar one

Bastian-Krause committed:
[WPE] Add support for touch based pointer events
https://bugs.webkit.org/show_bug.cgi?id=214870

Reviewed by Carlos Alberto Lopez Perez and Michael Catanzaro.

Currently all non-Apple ports support only pointer events of type
"mouse" because they a mapped 1:1 to mouse events. This commit adds
the pointer event support for touch events for the WPE port. With
this patch it should be easy for other ports to add pointer event support
as well.

The WPE port creates artificial motion, button down, button up events
for short touch gestures in PageClientImpl::doneWithTouchEvent() to keep
compatibility, see:

https://w3c.github.io/pointerevents/#compatibility-mapping-with-mouse-events

These NativeWebMouseEvents are undistinguishable from actual mouse
events and lead to additional pointer events of type mouse being
dispatched.
Fix this by passing the SyntheticClickType down to the WebMouseEvent and
PlatformMouseEvent. With this, WPE can make use of the code path
introduced in https://bugs.webkit.org/show_bug.cgi?id=205551 due to the
same issue.

Initial patch version by Marco Felsch <m.felsch@pengutronix.de>. Tested
and improved by Bastian Krause <bst@pengutronix.de>.

* Source/WebCore/SourcesWPE.txt: Add dom/wpe/PointerEventWPE.cpp, see
below.
* Source/WebCore/dom/PointerEvent.h: Enable various pointer event paths
for WPE, move buttonForType() and buttonsForType() generics from
PointerEventIOS.
* Source/WebCore/dom/wpe/PointerEventWPE.cpp: Adapted for WPE from
Source/WebCore/dom/ios/PointerEventIOS.cpp.
(WebCore::pointerEventType): Add a translation function similar to
PointerEventIOS's pointerEventType().
(WebCore::PointerEvent::create): Add create functions similar to
PointerEventIOS.
(WebCore::PointerEvent::PointerEvent):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleTouchEvent): Change point iteration to
make index value accessible. Add pointer target hit tests for
TouchReleased/TouchCancelled/TouchMoved. Implement WPE-specific
workaround to compensate for non-existent touch stationary events.
Finally dispatch pointer events.
* Source/WebCore/page/PointerCaptureController.cpp:
(WebCore::PointerCaptureController::cancelPointer): Enable existing
pointer event code paths for WPE.
* Source/WebCore/page/PointerCaptureController.h:
(WebCore::PointerCaptureController::CapturingData::hasAnyElement const):
* Source/WebCore/platform/PlatformTouchEvent.h:
(WebCore::PlatformTouchEvent::setTouchPoints): Allow setting touchPoints
for WPE to make the above mentioned touch stationary worarkound
possible.
* Source/WebCore/platform/PlatformTouchPoint.h:
(WebCore::PlatformTouchPoint::PlatformTouchPoint): Allow WPE to create
PlatformTouchPoints to make the above mentioned touch stationary
worarkound possible.
* Source/WebKit/Shared/NativeWebMouseEvent.h: Add synthetic click type.
* Source/WebKit/Shared/WebEventConversion.cpp:
(WebKit::WebKit2PlatformMouseEvent::WebKit2PlatformMouseEvent): Pass the
SyntheticClickType to the PlatformMouseEvent, for WPE only because effects
on other ports are unknown at this time.
* Source/WebKit/Shared/libwpe/NativeWebMouseEventLibWPE.cpp:
(WebKit::NativeWebMouseEvent::NativeWebMouseEvent): Add synthetic click
type.
* Source/WebKit/Shared/libwpe/WebEventFactory.cpp: Add synthetic click
type.
(WebKit::WebEventFactory::createWebMouseEvent):
* Source/WebKit/Shared/libwpe/WebEventFactory.h: Pass synthetic click
type to WebMouseEvent.
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::doneWithTouchEvent): Set synthetic click type
for synthetic mouse events to NativeWebMouseEvent.

Canonical link: https://commits.webkit.org/263969@main
